### DISCUSSION TITLE
Runout filament sensor

### DISCUSSION THREAD

**Discussion by juanvelo** [jschuh/klipper-macros#127] [Q&A] [answered]
Greetings
I'm trying to set up my runout filament sensor in Jschuh's macros and I can't get it to work.
Could someone give me a hand.
Thank you so much

**Comment by jschuh** [comment, marked as answer]
I keep my runout sensors positioned several inches before the extruder gears, so that when they trigger there's still enough filament available that the current layer will complete and the pause will run during the layer change:

```
[filament_switch_sensor runout]
pause_on_runout: False
switch_pin: <put your pin config here>
runout_gcode:
  # Pause immediately if we're in PRINT_START
  {% if printer.print_stats.info.current_layer < 1 %}
    PAUSE
  # Pause at the end of the layer if we're anywhere else
  {% elif printer.print_stats.info.current_layer <=
            printer.print_stats.info.total_layer and
          not printer.pause_resume.is_paused %}
    PAUSE_NEXT_LAYER
  {% endif %}
```

If you have a filament sensor that's very close to the extruder you probably want to pause right away, and may even want it to handle filament loading and unloading.

```
[filament_switch_sensor runout]
pin: <put your pin config here>
pause_on_runout: True
insert_gcode: # Remove this section if you don't want to automatically load on insertion.
  LOAD_FILAMENT
runout_gcode: # Remove this section if you want to pause without unloading.
  UNLOAD_FILAMENT
```

**Reply by juanvelo** [reply]
Thank you very much for answering.
I have tested it and it works correctly.
Excellent work!
